setup.py

```diff
@@ -18,13 +18,71 @@
 import setuptools
 from setuptools import setup
 from pathlib import Path
+from packaging import tags
+import os
+import json
+import requests
+import itertools
 import re
+from tqdm import tqdm
+
+
+def _get_paths(binaries, root_dir=""):
+    paths = []
+    if isinstance(binaries, str):
+        return [os.path.join(root_dir, binaries)]
+    elif isinstance(binaries, dict):
+        for k, v in binaries.items():
+            paths += _get_paths(v, os.path.join(root_dir, k))
+    else:
+        for i in binaries:
+            paths += _get_paths(i, root_dir)
+    return paths
 
 
 def _strip(line):
     return line.split(" ")[0].split("#")[0].split(",")[0]
 
 
+# Download all relevant binaries in binaries.json
+all_tags = list(tags.sys_tags())
+binaries = json.load(open("binaries.json"))
+paths = _get_paths(binaries)
+terminate = False
+pbar = None
+spinner = itertools.cycle(["-", "\\", "|", "/"])
+version = os.environ["VERSION"] if "VERSION" in os.environ else "main"
+print(f"Locating binaries {next(spinner)} ", end="")
+
+for tag in all_tags:
+    print(f"\rLocating binaries {next(spinner)} ", end="")
+    if terminate:
+        pbar.close()
+        break
+    for i, path in enumerate(paths):
+        if os.path.exists(path):
+            continue
+        folders = path.split(os.sep)
+        folder_path, file_path = os.sep.join(folders[:-1]), folders[-1]
+        file_name = f"{file_path[:-3]}_{tag}.so"
+        search_path = f"compiler/{file_name}"
+        r = requests.get(
+            f"https://github.com/unifyai/binaries/raw/{version}/{search_path}",
+            timeout=40,
+        )
+        if r.status_code == 200:
+            if pbar is None:
+                print()
+                print("Downloading binaries ...")
+                pbar = tqdm(total=len(paths))
+            with open(path, "wb") as f:
```

Comment:
can't pass over this point, I get:
```
          with open(path, "wb") as f:
      FileNotFoundError: [Errno 2] No such file or directory: 'ivy/compiler/utils/III.so'
```

Seems to work if I create the utils dir manually, so I'm assuming we are missing `os.makedirs(os.path.dirname(path), exist_ok=True)` before `open`

Reply:
done, thanks for pointing that out!